doc/developer/design/2024_10_08_expression_cache.md

```diff
@@ -0,0 +1,138 @@
+
+  - deployment generation
+  - object global ID
+  - expression type (local MIR, global MIR, LIR, etc)
+
+The value will be a serialized version of the optimized expression. An `environmentd` process with
+deploy generation `n`, will never be expected to look at a serialized expression with a deploy
+generation `m` s.t. `n != m`. Therefore, there are no forwards or backwards compatibility needed on
+the serialized representation of expressions. The cache will also be made  durable so that it's
+available after a restart, at least within the same deployment generation.
+
+Upgrading an environment will look something like this:
+
+1. Start deploy generation `n` in read-only mode.
+2. Populate the expression cache for generation `n`.
+3. Start deploy generation `n` in read-write mode.
+4. Read optimized expressions from cache.
+
+Restarting an environment will look something like this:
+
+1. Start deploy generation `n` in read-write mode.
+2. Read optimized expressions from cache.
+
+### Cache API
+
+Below is the API that the cache will present. It may be further wrapped with typed methods that
+take care of serializing and deserializing bytes. Additionally, we probably don't need a trait when
+implementing.
+
+```Rust
+trait ExpressionCache {
+    /// Returns the `expression_type` of `global_id` that is currently deployed in a cluster. This
+    /// will not change in-between restarts as result of DDL, as long as `global_id` exists.
+    fn get_deployed_expression(&self, global_id: GlobalId, expression_type: ExpressionType) -> Option<Bytes>;
+
+    /// Returns the `expression_type` of `global_id` based on the current catalog contents of
+    /// `deploy_generation`. This may change in-between restarts as result of DDL.
+    fn get_durable_expression(&self, deploy_generation: u64, global_id: GlobalId, expression_type: ExpressionType) -> Option<Bytes>;
+
+    /// Durably inserts `expression`, with key `(deploy_generation, global_id, expression_type)`.
+    ///
+    /// Panics if `(deploy_generation, global_id, expression_type)` already exists.
+    fn insert_expression(&mut self, deploy_generation: u64, global_id: GlobalId, expression_type: ExpressionType, expression: Bytes);
+
+    /// Durably remove and return all entries in `deploy_generation` that depend on an ID in
+    /// `dropped_ids`.
+    fn invalidate_entries(&mut self, deploy_generation: u64, dropped_ids: BTreeSet<GlobalId>) -> Vec<(GlobalId, ExpressionType)>;
+
+    /// Durably removes all entries in `deploy_generation`.
+    fn remove_deploy_generation(&mut self, deploy_generation: u64);
+
+    /// Remove all entries that depend on a global ID that is not present in `txn`.
+    fn reconcile(&mut self, txn: mz_catalog::durable::Transaction);
+}
+```
+
+### Startup
+
+Below is a detailed set of steps that will happen in startup.
+
+1. Call `ExpressionCache::reconcile` to remove any invalid entries.
+2. While opening the catalog, for each object:
```

Comment:
I'm a little confused by this. The design above---where `(generation, global_id, expr)` form a key that maps to a blob of optimized code---makes sense to drop in at various levels of calls to the `Optimizer`, which will, say, map HIR to one of the MIR newtypes, or MIR to LIR.

But what the design here seems like we'd be working at a higher level, skipping `Optimizer` entirely. In that case we could have a single, top-level cache from SQL straight to LIR or FlatPlan.

I think both designs are sensible. The former lets us make fewer changes; the latter reduces effort (if we have a cache hit from HIR to MIR, surely we can expect hits from the rest of the pipeline, too!).

Reply:
> But what the design here seems like we'd be working at a higher level, skipping `Optimizer` entirely. In that case we could have a single, top-level cache from SQL straight to LIR or FlatPlan.

Yes, as I understand it the plan is to skip the `Optimizer` entirely. But we still need these intermediate plans types because the catalog hangs on to the intermediate plans:

https://github.com/MaterializeInc/materialize/blob/bff231953f4bb97b70cae81bdd6dd1716dbf8cec/src/adapter/src/catalog.rs#L145-L151

I think for something with `EXPLAIN`.

Reply:
I see. So we could think of the cache as mapping `(generation, global_id, sql_expression)` to a tuple `(optimized_mir, physical_plan, metainfo, notices)`, then?

Reply:
Great! Another point in favor of the all-or-nothing approach: no implicit invariant around the validity of the generated intermediate expressions.

Reply:
I like the new approach better as well.